### DISCUSSION TITLE
Explicit function to convert to/from method parameters

### DISCUSSION THREAD

**Discussion by NewbiZ** [PyO3/pyo3#2774] [Questions] [answered]
I have a set of structs that have `rust_decimal::Decimal` as fields, as well as method parameters.

I would like to expose these structs to Python, but `rust_decimal::Decimal` is defined in an external crate, and `IntoPy<PyObject>` as well, so I have no clue as to how I could tell PyO3 how to convert to/from these structs.

**Comment by birkenfeld** [comment]
You would have to define a "newtype wrapper" struct like
```
struct Decimal(rust_decimal::Decimal);
```
and implement the `pyo3` traits for that.

Another way is to contribute integration to `pyo3`, which would make sense if we can agree that `rust_decimal` is "the most common" Rust decimal crate...

**Reply by NewbiZ** [reply]
I thought about creating a wrapper type, but to be honest I cannot justify all the amount of boilerplate it would require in my project. I would have to implement/forward all the zillions of traits that `rust_decimal::Decimal` provides for each operator and conversion, as well as all its specific methods.

At this point I am more keen to fork pyo3 and add `rust_decimal::Decimal` support, this seems like much less work (will contribute the PR in the original repo as well of course).

I'm just thinking I cannot be the only one to have this issue, and there are already quite a number of forks that I found created just for this very purpose.

I wondered if it would not be more flexible for PyO3 to provide a way of explicitely providing a conversion function for a specific parameter/type to be used instead of `IntoPy<PyObject>::into_py`, in the same vein as what `serde` does for custom serialization/deserialization of a field.
Something that would look like:

```rust
fn into_py_custom_foreign_type(self, py: Python<'_>) -> PyObject {
    [...]
}

#[pyfunction(convert = [b => into_py_custom_foreign_type])]
fn foo(a: u32, b: CustomForeignType) {
    [...]
}
```

**Reply by davidhewitt** [reply]
It looks like there is both `rust_decimal` and `bigdecimal` with roughly equal usage (`rust_decimal` just ahead). `rust_decimal` looks more feature-rich and better maintained, so I think arguments could be made to either support just `rust_decimal` based on maintenance or support both based on usage (e.g. we support both `anyhow` and `eyre` for error reporting, and I am open to us considering supporting `time` to add alongside `chrono`).

That said, the `Decimal` type in Python is not exposed via the C API as far as I'm aware. So far we have generally avoided adding support in PyO3 for such things. That's not to say we can't do it, but a discussion would be needed to consider where the scope of such support should fall.

On your original question, note that you can also annotate b with [`#[pyo3(from_py_with = "into_py_custom_foreign_type")]`](https://pyo3.rs/v0.17.3/function.html?highlight=from_py_with#per-argument-options), which is very similar to your proposed "convert".

**Reply by NewbiZ** [reply]
> On your original question, note that you can also annotate b with #[pyo3(from_py_with = "into_py_custom_foreign_type")], which is very similar to your proposed "convert".

Oh that seems great, I did not know this macro. Is there an equivalent for returned types?

**Reply by davidhewitt** [reply]
There isn't, because it's already straightforward to just wrap return values with either a wrapper type or a function call (as you prefer).

(The advantage for having the macro for the inputs is that we can then in the PyO3 framework give hints to the caller about which function input failed to convert.)

**Comment by cardoe** [comment]
I've made a rough swag at adding this into pyo3 proper at https://github.com/cardoe/pyo3/tree/decimal right now there's no docs and the tests aren't great. I'm hoping someone like @davidhewitt can tell me if this approach is okay and I'll work at getting this PR worthy.

**Reply by adamreichold** [reply]
I think you would want cache the `decimal.Decimal` constructor using a [`GILOnceCell`](https://docs.rs/pyo3/latest/pyo3/once_cell/struct.GILOnceCell.html). I also think using [`proptest`](https://docs.rs/proptest/latest/proptest/) would be nice to expand on the roundtrip checking (as we already do in other conversion modules).

**Reply by adamreichold** [reply]
(Also note that `Decimal::from_str_exact` does not seem to be present in `rust-decimal` 1.0, so a higher minimum version should probably be required.)

**Reply by cardoe** [reply]
Thanks for the suggestions. I'll incorporate them.

**Reply by cardoe** [reply]
Do you have any suggestions for the TODO items by any chance?

**Reply by adamreichold** [reply]
> should this check that the underlying type is of decimal.Decimal ?

I would advise against that as `FromPyObject` is not a downcast but rather something like a "conversion constructor", i.e. it would be good if it works with an integer in addition a `decimal.Decimal`.

> should this use obj.call_method0(intern!(py, "__str__"))?

I think our `PyAny::str` calling `PyObject_Str` should be equivalent to that but simpler and potentially faster as well.

> how do we handle the error gracefully?

IIRC this is an open issue, meaning that `ToPyObject` will need to become fallible for other conversions as well but we have not yet worked out a sufficiently ergonomic way forward on this meaning that you need to stick with the panics for now.

> is PyTypeError the right thing?

I think that does depend on the underlying error condition but in this case I would generally gravitate more towards `PyValueError` if you do not decide to check for the value having the `decimal.Decimal` type.

**Reply by davidhewitt** [reply]
I think it's easier to discuss the TODO items as a PR (your existing commit would already be fine to open a PR for discussion).

I'm a little uneasy about the performance characteristics of this, you may want to benchmark it for your use case. It could be worth also asking upstream CPython about the feasibility of adding a C API for this, which might help us reduce conversion cost? (Speculating wildly, I'm not sure how interoperable the C representation actually is.)

**Comment by mejrs** [comment]
> I thought about creating a wrapper type, but to be honest I cannot justify all the amount of boilerplate it would require in my project.

I think this boilerplate is much less than you expect, so I'd still lean towards this over providing the conversions ourselves. IMO we should only do conversions if we can do it in a "good" way, i.e. if there's a C api we can call. decimal has no such C api.

**Comment by cardoe** [comment, marked as answer]
@NewbiZ you might want to mark this as answered. #3016 has landed which adds support natively to pyo3 for Decimal via rust_decimal.

**Reply by cardoe** [reply]
Now part of the 0.19.0 release.
